nmxact/xact/image.go

```diff
@@ -185,33 +204,149 @@ func nextImageUploadReq(s sesn.Sesn, upgrade bool, data []byte, off int, imageNu
 	return r, nil
 }
 
-func (c *ImageUploadCmd) Run(s sesn.Sesn) (Result, error) {
-	res := newImageUploadResult()
+func (t *ImageUploadIntTracker) update(t_off int, r_off int, status int) {
```

Comment:
Something about this doesn't seem right to me.  It is very possible that I'm just confusing myself here, so please don't hold back if I'm out to lunch.

* `t_off` is the offset of the request we sent right before we started listening.
* `r_off` is the offset of the response we just received.

Chances are these offsets have no relation to one another.  That is, the response does not correspond to this particular request.  So why does this function accept these values as a pair?  To me, sending and receiving have different effects on the map, so they should be processed separately.

I would expect something like this:
ON TRANSMIT:
1. Indicate that we are expecting a response to the request we just transmitted (set an element in the map).

ON RECEIVE:
1. We received a successful response with offset=x.
2. Are we expecting this request (look in map)
a. Yes: delete entry from map.
b. No: ignore.

I don't understand the counts being stored in the map.  Like I said - it may totally make sense.

Reply:
The `update`  function: It is a generic one. It is called every time an error response or response requesting a chunk is received. Since the response handling is a goroutine, it knows what offset was sent. The response offset is not necessarily related to the transmit offset. That is right. As the tracker is being modified under the mutex lock, I was trying to minimize access.

Your approach makes sense to me. I think I used the map thinking the response offsets can be random order, which probably is not true. At least in my logs I don't see that. 
Probably just one repeated offset request is indication enough that a particular chunk@offset needs re-transmission and the map may not be needed.

Reply:
I have reworked the logic to have just one loop and simply block when there is no room in the window and wait for more. 
Use a `done` channel to indicate when all the offsets are received by the remote side.
Tests fine with Serial and BT transport.

Reply:
I "feel" that there could be a possibility that the offsets in responses repeat e.g 0, 357, 357, 357, 750, 1120 etc. Hence keeping the count in the maps around for now. Based on more testing, we can potentially get rid of it.